File: Python/36.Valid Sudoku.py
#判断一个数独题目的初始状态是否正确(每行每列每个小正方形有没有重复数字)
class Solution(object):
    def isValidSudoku(self, board):
        across,vertical,square= [([] * 9) for i in range(9)],[([] * 9) for i in range(9)], [([] * 9) for i in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    continue
                if board[i][j] in across[i]:
                    return False
                else:
                    across[i].append(board[i][j])
                if board[i][j] in vertical[j]:
                    return False
                else:
                    vertical[j].append(board[i][j])
                if board[i][j] in square[(i//3)*3+j//3]:
                    return False
                else:
                    square[(i//3)*3+j//3].append(board[i][j])
        return True

s = Solution()
print(s.isValidSudoku([['5','3','.','.','7','.','.','.','.'],
                       ['6','.','.','1','9','5','.','.','.'],
                       ['.','9','8','.','.','.','.','6','.'],
                       ['8','.','.','.','6','.','.','.','3'],
                       ['4','.','.','8','.','3','.','.','1'],
                       ['7','.','.','.','2','.','.','.','6'],
                       ['.','6','.','.','.','.','2','8','.'],
                       ['.','.','.','4','1','9','.','.','5'],
                       ['.','.','.','.','8','.','.','7','9']]))

File: C++/Hash Table/36.Valid Sudoku.cpp
/*
判断一个数独是否是合法的。
*/

#include <iostream>
#include <cstring>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
	bool isValidSudoku(vector<string>& board) {
		vector<unordered_map<char, int>>	row(9);
		vector<unordered_map<char, int>>	col(9);
		vector<unordered_map<char, int>>	square(9);
		for (int i = 0; i < 9; ++i)
		{
			for (int j = 0; j < 9; ++j)
			{
				if (board[i][j] == '.')	continue;
				if (row[i].find(board[i][j]) == row[i].end())
					row[i].insert({ board[i][j], 1 });
				else
					return false;

				if (col[j].find(board[i][j]) == col[j].end())
					col[j].insert({ board[i][j], 1 });
				else
					return false;

				if (square[i / 3 * 3 + j / 3].find(board[i][j]) == square[i / 3 * 3 + j / 3].end())
					square[i / 3*3 + j / 3].insert({ board[i][j], 1 });
				else
					return false;
			}
		}
		return true;
	}

};

int main()
{
	Solution s;
	vector<string> board{ { ".87654321" }, { "2........" }, { "3........" }, { "4........" }, { "5........" }, { "6........" }, { "7........" }, { "8........" }, { "9........" } };
	cout << s.isValidSudoku(board) << endl;
}

